### DISCUSSION TITLE
Could subscriptions return an error?

### DISCUSSION THREAD

**Discussion by ricardo-valero** [gqty-dev/gqty#1283] [Ideas]
Would it be possible to have an api similar to `useMutation` for `useSubscription`?

```ts
  const [sub, subState] = useSubscription(
    (sub, args: MyArgs) => sub.mySub(args),
    {
      onData: data => console.log(data),
      onError: error => console.log(error),
    },
  )
```

**Comment by vicary** [comment]
`useSubscription()` currently takes no arguments, it should be trivial to expose scheduler errors via callback.

Subscription errors are mostly network issues, which is usually silently handled via reconnection. I have yet to see a server implementation explicitly support custom errors wiring down the WebSocket channel.

Is there any particular errors you are interested in handling this way?

**Reply by ricardo-valero** [reply]
My specific use case is that I have multiple sensors (using tcp sockets), so my subscription requires some args (ip address and port) and what I'd like is that if the connection fails (in my graphql server) then I can return the error, and then add some logic (retry, cancel, new args, etc)

**Reply by vicary** [reply]
Exposing [core subscription options](https://gqty.dev/docs/client/subscriptions#more-options) could be a trivial start, I can also see that current core callbacks could use more event coverage in the next major.

You should be able to add your logics via `failedReconnectionCallback` by setting `maxReconnectAttempts` to 0, so it immediately fails and you get to decide if new args should be applied.

Let me know if that's enough for you, I'll then make a minor release shortly.
